### PR TITLE
Unit test for index operation to get notification event

### PR DESCRIPTION
**Pull request by kavithacm**
### Description

- Unit test to check if it gets single notification event, given an index id in the Notification Event index.
- Checks if it creates a new index event and fetches it, if the given index id isn't present.
- Mocked all dependencies required by the function using Mockito and Mockk

All test in the NotificationEventIndexTests file successfully builds and all the test cases passed in local machine but it fails here due to some github action, because of which the PR couldn’t be merged currently. Once its resolved, the PR will be merged.

The solution to the above issue along with some insights has been created as a new issue and linked along with this PR.
https://github.com/opensearch-project/notifications/issues/319
 
### Issues Resolved
Add unit tests for index operation to get single event #96 
Linked the issue in the "linked issues" tab
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by zhongnansu** [comment]
Please submit the PR to `main` branch , as across opensearch plugins, we are following the same branching rules

**Comment by kavithacm** [comment]
Should i cancel this PR and create a new one in main branch?

**Comment by zhongnansu** [comment]
@kavithacm 
1. Make sure the github actions can pass, currently I see one is failing
2. Add at least 2 reviewers to your PR

**Comment by kavithacm** [comment]
> @kavithacm
> 
> 1. Make sure the github actions can pass, currently I see one is failing
> 2. Add at least 2 reviewers to your PR
> 
> some general [reference](https://github.com/opensearch-project/opensearch-plugins#contributing) about developing/contributing to a opensearch plugin project

1. Currently on it. The build was successful in my local machine. I'm facing with the "unapproved license" issue during the build in github
2. Can i add anyone for reviewing from our team?

**Comment by davidcui1225** [comment]
Can you add some details in the PR description about what this test case checks for, as well as rename the PR to a more descriptive name?

**Comment by zhongnansu** [comment]
> > @kavithacm
> > 
> > 1. Make sure the github actions can pass, currently I see one is failing
> > 2. Add at least 2 reviewers to your PR
> > 
> > some general [reference](https://github.com/opensearch-project/opensearch-plugins#contributing) about developing/contributing to a opensearch plugin project
> 
> 1. Currently on it. The build was successful in my local machine. I'm facing with the "unapproved license" issue during the build in github
> 2. Can i add anyone for reviewing from our team?

1. this could happen sometimes as the github action has its own build env

**Comment by kavithacm** [comment]
> Can you add some details in the PR description about what this test case checks for, as well as rename the PR to a more descriptive name?

I had changed it. Will this do?

**Comment by zhongnansu** [comment]
Also link the issue that you are resolving in this PR desciption, if there's not an existing issue, create one.

**Comment by kavithacm** [comment]
> > > @kavithacm
> > > 
> > > 1. Make sure the github actions can pass, currently I see one is failing
> > > 2. Add at least 2 reviewers to your PR
> > > 
> > > some general [reference](https://github.com/opensearch-project/opensearch-plugins#contributing) about developing/contributing to a opensearch plugin project
> > 
> > 
> > 
> > 1. Currently on it. The build was successful in my local machine. I'm facing with the "unapproved license" issue during the build in github
> > 2. Can i add anyone for reviewing from our team?
> 
> 1. this could happen sometimes as the github action has its own build env

I fixed it. just the license header was missing

**Comment by kavithacm** [comment]
> > > Can you add some details in the PR description about what this test case checks for, as well as rename the PR to a more descriptive name?
> > 
> > 
> > I had changed it. Will this do?
> 
> @kavithacm Also link the issue that you are resolving in this PR desciption, if there's not an existing issue, create one.

Linked both the epic and the issue.

**Comment by chloe-zh** [comment]
Could you add a couple of test for the failure cases?

**Comment by kavithacm** [comment]
> Could you add a couple of test for the failure cases?

I have added some more test for failure cases. Will this do?

File: notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        mockkObject(NotificationEventDoc)
+        every { NotificationEventDoc.parse(any()) } returns eventDoc
+
+        val actualEventDocInfo = NotificationEventIndex.getNotificationEvent(id)
+        assertEquals(expectedEventDocInfo, actualEventDocInfo)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    @Test
+    fun `NotificationEventIndex should safely return null if response source is null`() {
```
[the remark this answers is not in the document]
Comment (zhongnansu) [reply notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt:133]: @kavithacm 
I think you misunderstood the concept of "index" and "document id" in OpenSearch. An index is a collection of documents. You have to create an index before you add any documents to it.(Each document has an id). In our case, notificationEvent Index needs to be initialized before we can add any "event" to it.

File: notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        val mockActionGet = mock(CreateIndexResponse::class.java)
+        mockCreateIndex(mockActionGet)
+        Assertions.assertThrows(IllegalStateException::class.java) {
+            NotificationEventIndex.getNotificationEvent(id)
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    @Test
+    fun `NotificationEventIndex should throw exception if index couldn't be created`() {
```
Comment (zhongnansu) [inline notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt:177]: Why an index couldn't be created and how is it achieved by your `mockIsIndexExists` method?
"Index already exists" shouldn't be a reason for `notificaitonGetSingleEvent` method to fail.

Comment (kavithacm) [reply notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt:177]: The function does't fail if the index already exists, it fails only if a new index couldn’t be created and the **mockIsIndexExist** is just to mock the original method **isIndexExists** which checks if the Index exist, prior to creating a new one. If the index exist already, it'll just fetch the notification event.

**mockIsIndexExist** is just a helper function to check the condition and the function **createIndex** (called by getNotificationEvent) is the one that causes the exception after it passes the **isIndexExists** function call.

File: notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+import org.opensearch.commons.notifications.model.NotificationEvent
+import org.opensearch.commons.notifications.model.SeverityType
+import org.opensearch.notifications.model.DocInfo
+import org.opensearch.notifications.model.DocMetadata
+import org.opensearch.notifications.model.NotificationEventDoc
+import org.opensearch.notifications.model.NotificationEventDocInfo
+import org.opensearch.threadpool.ThreadPool
+import java.time.Instant
+
+internal class NotificationEventIndexTest {
```
Comment (zhongnansu) [inline notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt:50]: Please follow the naming convention to name test classes and files as `***Tests`,  plural form

Comment (kavithacm) [reply notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt:50]: Sure i'll rename it now

File: notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt (hunk 2 above)
Comment (zhongnansu) [reply notifications/notifications/src/test/kotlin/org/opensearch/notifications/index/NotificationEventIndexTest.kt]: I think the exception thrown is actually NPE based on your code and your mock. I still don't think it's a valid test case to include, unless you can clarify that why an index creation could thrown exception. Or you mock that specific exceptions. Currently I see you use a general exception class to do the verification

**Comment by zhongnansu** [comment]
@kavithacm any status update of this PR?

**Comment by kavithacm** [comment]
> @kavithacm any status update of this PR?

I have created a new issue as suggested along with some insights on the error log and the modification to the build file and I've linked it to this PR.

**Pull request merged**